packages/typescript-estree/src/parseSettings/getProjectConfigFiles.ts

```diff
@@ -0,0 +1,80 @@
+import debug from 'debug';
+import fs from 'fs';
+import path from 'path';
+
+import type { ParseSettings } from '.';
+
+const log = debug('typescript-eslint:typescript-estree:getProjectConfigFiles');
+
+interface TSConfigMatch {
+  exists: boolean;
+  timestamp: number;
+}
+
+/**
+ * How many milliseconds we will respect a cache for checking an fs.existsSync
+ * check for a file file on disk.
+ */
+const RECHECK_FILE_THRESHOLD_MS = 50;
+
+const tsconfigMatchCache = new Map<string, TSConfigMatch>();
+
+/**
+ * Checks for a file file on disk, respecting a limited-time cache.
+ * This keeps a caches checked paths with existence and `Date.now()` timestamp.
+ * After `RECHECK_FILE_THRESHOLD_MS`, cache entries are ignored.
+ *
+ * @param filePath File path to check for existence on disk.
+ * @returns Whether the file exists on disk.
+ * @remarks
+ * We don't (yet!) have a way to attach file watchers on disk, but still need to
+ * cache file checks for rapid subsequent calls to fs.existsSync. See discussion
+ * in https://github.com/typescript-eslint/typescript-eslint/issues/101.
+ */
+function existsSyncCached(filePath: string): boolean {
+  const cached = tsconfigMatchCache.get(filePath);
+  const now = Date.now();
+
+  if (cached && now - cached.timestamp < RECHECK_FILE_THRESHOLD_MS) {
+    return cached.exists;
+  }
+
+  const exists = fs.existsSync(filePath);
+
+  tsconfigMatchCache.set(filePath, {
+    exists,
+    timestamp: now,
+  });
+
+  return exists;
+}
+
+export function getProjectConfigFiles(
+  parseSettings: ParseSettings,
+  projects: string | string[] | true | undefined,
+): string | string[] | undefined {
+  if (projects !== true) {
+    return projects;
+  }
+
+  log('Looking for tsconfig.json at or above file: %s', parseSettings.filePath);
+  let directory = path.dirname(parseSettings.filePath);
+
+  do {
+    const tsconfigPath = path.join(directory, 'tsconfig.json');
+    log('Checking tsconfig.json path: %s', tsconfigPath);
+
+    if (existsSyncCached(tsconfigPath)) {
+      return [tsconfigPath];
```

Comment:
In theory, this could also check whether the `tsconfig.json` includes `parseSettings.filePath` as described in https://github.com/typescript-eslint/typescript-eslint/issues/101#issuecomment-1327052709, and skip if it doesn't. But what if the user _wanted_ a particular `tsconfig.json` to include the file, and instead got some differently configured higher-in-the-project `tsconfig.json`? 😱 

In other words we have two cases to handle:

* 🍎 Closest `tsconfig.json` is matched but does not include the file
* 🍌 Closest `tsconfig.json` matches but does not include the file

I think we should stick with this simpler logic and not parse the `tsconfig.json` because the error message for 🍌 is quite informative as of #5762.

Reply:
I pinged a couple of folks, we'll see if anybody picks up. In the meantime I'd vote to merge this into `main`, and make a breaking change in `v6` if we really need. I'm leaning more strongly towards _explicitly break if misconfigured_ in general.